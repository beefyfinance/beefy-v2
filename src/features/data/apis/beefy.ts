import axios, { AxiosInstance } from 'axios';
import BigNumber from 'bignumber.js';
import { isString } from 'lodash';
import { ChainEntity } from '../entities/chain';
import { TokenEntity } from '../entities/token';
import { VaultEntity } from '../entities/vault';
import { mapValuesDeep } from '../utils/array-utils';
import { featureFlag_simulateBeefyApiError } from '../utils/feature-flags';

interface ApyGovVault {
  vaultApr: number;
}
interface ApyMaxiVault {
  totalApy: number;
}
interface ApyStandard {
  beefyPerformanceFee: number;
  vaultApr: number;
  compoundingsPerYear: number;
  vaultApy: number;
  tradingApr?: number;
  totalApy: number;
  // todo: does it make sense to have fees and apy in the same entities?
  lpFee: number;
}
export type ApyData = ApyGovVault | ApyMaxiVault | ApyStandard;
export function isStandardVaultApy(apy: ApyData): apy is ApyStandard {
  return 'compoundingsPerYear' in apy;
}
export function isGovVaultApy(apy: ApyData): apy is ApyGovVault {
  return 'vaultApr' in apy && !('compoundingsPerYear' in apy);
}
export function isMaxiVaultApy(apy: ApyData): apy is ApyMaxiVault {
  return 'totalApy' in apy && !('compoundingsPerYear' in apy);
}

export interface BeefyAPITokenPricesResponse {
  [tokenId: TokenEntity['id']]: number;
}
export interface BeefyAPIBreakdownResponse {
  [vaultId: VaultEntity['id']]: ApyData;
}

export interface BeefyAPIBuybackResponse {
  // those are of type string but they represent numbers
  [chainId: ChainEntity['id']]: { buybackTokenAmount: BigNumber; buybackUsdAmount: BigNumber };
}

// note that there is more infos but we don't need it
type BeefyAPIVaultsResponse = {
  id: string;
  lastHarvest?: number | string;
}[];

export class BeefyAPI {
  public api: AxiosInstance;
  public data: AxiosInstance;

  constructor() {
    // this could be mocked by passing mock axios to the constructor
    this.api = axios.create({
      baseURL: 'https://api.beefy.finance',
      timeout: 30 * 1000,
    });
    this.data = axios.create({
      baseURL: 'https://data.beefy.finance',
      timeout: 30 * 1000,
    });
  }

  // here we can nicely type the responses
  public async getPrices(): Promise<BeefyAPITokenPricesResponse> {
    if (featureFlag_simulateBeefyApiError('prices')) {
      throw new Error('Simulated beefy api error');
    }

    const res = await this.api.get('/prices', { params: { _: this.getCacheBuster('hour') } });
    return res.data;
  }

  // i'm not 100% certain about the return type
  // are those token ids ?
  public async getLPs(): Promise<BeefyAPITokenPricesResponse> {
    if (featureFlag_simulateBeefyApiError('lps')) {
      throw new Error('Simulated beefy api error');
    }

    const res = await this.api.get('/lps', { params: { _: this.getCacheBuster('hour') } });
    return res.data;
  }

  public async getBreakdown(): Promise<BeefyAPIBreakdownResponse> {
    if (featureFlag_simulateBeefyApiError('apy')) {
      throw new Error('Simulated beefy api error');
    }

    const res = await this.api.get('/apy/breakdown', {
      params: { _: this.getCacheBuster('day') },
    });

    // somehow, all vaultApr are currently strings, we need to fix that before sending
    // the data to be processed
    const data = mapValuesDeep(res.data, (val, key) => {
      if (key === 'vaultApr' && isString(val)) {
        val = parseFloat(val);
      }
      return val;
    });

    return data;
  }

  public async getBuyBack(): Promise<BeefyAPIBuybackResponse> {
    if (featureFlag_simulateBeefyApiError('buyback')) {
      throw new Error('Simulated beefy api error');
    }

    type ResponseType = {
      data: {
        [chainId: ChainEntity['id']]: {
          buybackTokenAmount: string;
          buybackUsdAmount: string;
        };
      };
    };
    const strRes = await this.api.get<ResponseType>('/bifibuyback', {
      params: { _: this.getCacheBuster('hour') },
    });
    // format result with big number as we get strings from the api
    const res: BeefyAPIBuybackResponse = {};
    for (const chainId in strRes.data) {
      res[chainId] = {
        buybackTokenAmount: new BigNumber(strRes.data[chainId].buybackTokenAmount),
        buybackUsdAmount: new BigNumber(strRes.data[chainId].buybackUsdAmount),
      };
    }
    return res;
  }

  /**
   * For now we fetch lastHarvest from the api
   * TODO: fetch this from the contract directly
   */
  public async getVaultLastHarvest(vaultId: VaultEntity['id']): Promise<null | Date> {
    const res = await this.api.get<BeefyAPIVaultsResponse>('/vaults', {
      params: { _: this.getCacheBuster('hour') },
    });

    // const vault = data.filter(vault => vault.id.includes(vaultId));
    const vaultConfig = res.data.find(vault => {
      return vault.id === vaultId;
    });

    if (!vaultConfig) {
      // vault is not harvestable (most probably a gov vault)
      return null;
    }
    if (!('lastHarvest' in vaultConfig)) {
      return null;
    }
    let lh = 0;
    if (isString(vaultConfig.lastHarvest)) {
      lh = parseInt(vaultConfig.lastHarvest);
    } else {
      lh = vaultConfig.lastHarvest;
    }
    if (lh === 0) {
      return null;
    }

    return new Date(lh * 1000);
  }

  // maybe have a local cache instead of this cache busting
  // have to check if this returns browser cache before doing so
  protected getCacheBuster(mode: 'hour' | 'day' = 'day'): number {
    if (mode === 'hour') {
      return Math.trunc(Date.now() / (1000 * 60));
    } else {
      const cache = new Date();
      cache.setMinutes(0, 0, 0);
      return cache.getTime();
    }
  }
}
