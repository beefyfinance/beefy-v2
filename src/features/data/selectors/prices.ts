import { createGlobalDataSelector, hasLoaderFulfilledOnce } from './data-loader-helpers.ts';

export const selectIsPricesAvailable = createGlobalDataSelector('prices', hasLoaderFulfilledOnce);
