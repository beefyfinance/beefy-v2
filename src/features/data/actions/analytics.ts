import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import { getAnalyticsApi } from '../apis/instances';
import type { VaultTimelineAnalyticsEntity } from '../entities/analytics';
import BigNumber from 'bignumber.js';
import type { AnalyticsPriceResponse, TimeBucketType } from '../apis/analytics/analytics-types';
import type { VaultEntity } from '../entities/vault';
import { isFiniteNumber } from '../../../helpers/number';

export interface fetchWalletTimelineFulfilled {
  timeline: VaultTimelineAnalyticsEntity[];
  walletAddress: string;
  state: BeefyState;
}

export const fetchWalletTimeline = createAsyncThunk<
  fetchWalletTimelineFulfilled,
  { address: string },
  { state: BeefyState }
>('analytics/fetchWalletTimeline', async ({ address }, { getState }) => {
  const api = await getAnalyticsApi();

  const userTimeline = await api.getWalletTimeline(address);

  const timeline = userTimeline.map(row => {
    return {
      chain: row.chain,
      datetime: new Date(row.datetime),
      displayName: row.display_name,
      isEol: row.is_eol,
      productKey: row.product_key,
      shareBalance: new BigNumber(row.share_balance),
      shareDiff: new BigNumber(row.share_diff),
      shareToUnderlyingPrice: new BigNumber(row.share_to_underlying_price),
      underlyingBalance: new BigNumber(row.underlying_balance),
      underlyingDiff: new BigNumber(row.underlying_diff),
      underlyingToUsdPrice: isFiniteNumber(row.underlying_to_usd_price)
        ? new BigNumber(row.underlying_to_usd_price)
        : null,
      usdBalance: isFiniteNumber(row.usd_balance) ? new BigNumber(row.usd_balance) : null,
      usdDiff: isFiniteNumber(row.usd_diff) ? new BigNumber(row.usd_diff) : null,
    };
  });

  return { timeline, walletAddress: address.toLowerCase(), state: getState() };
});

interface DataMartPricesFulfilled {
  data: AnalyticsPriceResponse;
  vaultId: VaultEntity['id'];
  timebucket: TimeBucketType;
  walletAddress: string;
  state: BeefyState;
}

interface DataMartPricesProps {
  productKey: string;
  timebucket: TimeBucketType;
  walletAddress: string;
  vaultId: VaultEntity['id'];
}

export const fetchShareToUnderlying = createAsyncThunk<
  DataMartPricesFulfilled,
  DataMartPricesProps,
  { state: BeefyState }
>(
  'analytics/fetchShareToUnderlying',
  async ({ productKey, walletAddress, timebucket, vaultId }, { getState }) => {
    const api = await getAnalyticsApi();
    const data = await api.getVaultPrices(productKey, 'share_to_underlying', timebucket);
    return {
      data,
      vaultId,
      timebucket,
      walletAddress: walletAddress.toLocaleLowerCase(),
      state: getState(),
    };
  }
);

export const fetchUnderlyingToUsd = createAsyncThunk<
  DataMartPricesFulfilled,
  DataMartPricesProps,
  { state: BeefyState }
>(
  'analytics/fetchUnderlyingToUsd',
  async ({ productKey, timebucket, walletAddress, vaultId }, { getState }) => {
    const api = await getAnalyticsApi();
    const data = await api.getVaultPrices(productKey, 'underlying_to_usd', timebucket);
    return {
      data,
      vaultId,
      timebucket,
      walletAddress: walletAddress.toLocaleLowerCase(),
      state: getState(),
    };
  }
);
