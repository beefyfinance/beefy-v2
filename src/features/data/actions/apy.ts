import { createAsyncThunk } from '@reduxjs/toolkit';
import { BeefyAPIBreakdownResponse, BeefyAPIHistoricalAPYResponse } from '../apis/beefy';
import { getBeefyApi } from '../apis/instances';

export const fetchApyAction = createAsyncThunk<BeefyAPIBreakdownResponse, {}>(
  'prices/fetchApy',
  async () => {
    const api = getBeefyApi();
    const prices = await api.getBreakdown();
    return prices;
  }
);

export const fetchHistoricalApy = createAsyncThunk<BeefyAPIHistoricalAPYResponse, {}>(
  'prices/fetchHistoricalApy',
  async () => {
    const api = getBeefyApi();
    const prices = await api.getHistoricalAPY();
    return prices;
  }
);
