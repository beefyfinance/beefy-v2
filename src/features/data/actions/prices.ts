import { createAsyncThunk } from '@reduxjs/toolkit';
import { BeefyAPITokenPricesResponse } from '../apis/beefy';
import { getBeefyApi } from '../apis/instances';

export const fetchAllPricesAction = createAsyncThunk<BeefyAPITokenPricesResponse, {}>(
  'prices/fetchAllPricesAction',
  async () => {
    const api = getBeefyApi();
    const [prices, lpPrices] = await Promise.all([api.getPrices(), api.getLPs()]);
    return {
      ...prices,
      ...lpPrices,
    };
  }
);
