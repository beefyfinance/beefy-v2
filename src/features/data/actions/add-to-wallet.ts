import type { ChainEntity } from '../entities/chain';
import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { TokenEntity } from '../entities/token';
import { selectTokenByAddress } from '../selectors/tokens';
import { getSingleAssetSrc } from '../../../helpers/singleAssetSrc';

export type AddToWalletParams = {
  chainId: ChainEntity['id'];
  tokenAddress: TokenEntity['address'];
  customIconUrl?: string;
};

export type AddTokenToWalletPayload = {
  token: TokenEntity;
  iconUrl: string;
};

function getTokenIconUrl(
  id: TokenEntity['id'],
  chainId: TokenEntity['chainId'],
  customUrl?: string
): string {
  if (customUrl) {
    return customUrl;
  }

  const assetSrc = getSingleAssetSrc(id, chainId);
  if (assetSrc) {
    return `${window.location.origin}${assetSrc}`;
  }

  return '';
}

export const addTokenToWalletAction = createAsyncThunk<
  AddTokenToWalletPayload,
  AddToWalletParams,
  { state: BeefyState }
>('addToWallet/open', async ({ chainId, tokenAddress, customIconUrl }, { getState }) => {
  const state = getState();
  const token = selectTokenByAddress(state, chainId, tokenAddress);
  const iconUrl = getTokenIconUrl(token.id, token.chainId, customIconUrl);

  return {
    token,
    iconUrl,
  };
});
