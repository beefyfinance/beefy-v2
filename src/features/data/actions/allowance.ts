import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { FetchAllAllowanceResult } from '../apis/allowance/allowance-types';
import { getAllowanceApi } from '../apis/instances';
import type { ChainEntity } from '../entities/chain';
import type { TokenErc20 } from '../entities/token';
import type { VaultGov, VaultStandard } from '../entities/vault';
import { isGovVault } from '../entities/vault';
import { selectBoostById, selectBoostsByChainId } from '../selectors/boosts';
import { selectChainById } from '../selectors/chains';
import { selectVaultByChainId, selectVaultById } from '../selectors/vaults';
import { selectWalletAddress } from '../selectors/wallet';

interface ActionParams {
  chainId: ChainEntity['id'];
  walletAddress: string;
}

export interface FetchAllAllowanceFulfilledPayload {
  chainId: ChainEntity['id'];
  data: FetchAllAllowanceResult;
}

export const fetchAllAllowanceAction = createAsyncThunk<
  FetchAllAllowanceFulfilledPayload,
  ActionParams,
  { state: BeefyState }
>('allowance/fetchAllAllowanceAction', async ({ chainId, walletAddress }, { getState }) => {
  const state = getState();

  const userAddress = walletAddress || selectWalletAddress(state);
  const chain = selectChainById(state, chainId);
  const api = await getAllowanceApi(chain);

  // maybe have a way to retrieve those easily
  const boosts = selectBoostsByChainId(state, chainId).map(vaultId =>
    selectBoostById(state, vaultId)
  );
  const allVaults = selectVaultByChainId(state, chainId).map(vaultId =>
    selectVaultById(state, vaultId)
  );
  const standardVaults: VaultStandard[] = [];
  const govVaults: VaultGov[] = [];
  for (const vault of allVaults) {
    if (isGovVault(vault)) {
      govVaults.push(vault);
    } else {
      standardVaults.push(vault);
    }
  }

  // always re-fetch state as late as possible
  const data = await api.fetchAllAllowances(
    // always re-fetch state as late as possible
    getState(),
    standardVaults,
    govVaults,
    boosts,
    userAddress
  );
  return { chainId, data };
});

interface FetchAllowanceActionParams {
  chainId: ChainEntity['id'];
  spenderAddress: string;
  tokens: TokenErc20[];
  walletAddress: string;
}

export const fetchAllowanceAction = createAsyncThunk<
  FetchAllAllowanceFulfilledPayload,
  FetchAllowanceActionParams,
  { state: BeefyState }
>(
  'allowance/fetchAllowanceAction',
  async ({ chainId, spenderAddress, tokens, walletAddress }, { getState }) => {
    const state = getState();
    const userAddress = walletAddress || selectWalletAddress(state);
    const chain = selectChainById(state, chainId);
    const api = await getAllowanceApi(chain);

    const allowanceRes =
      userAddress && spenderAddress
        ? await api.fetchTokensAllowance(getState(), tokens, userAddress, spenderAddress)
        : [];

    return { chainId, data: allowanceRes };
  }
);
