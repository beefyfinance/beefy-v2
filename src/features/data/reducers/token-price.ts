import { createSlice } from '@reduxjs/toolkit';
import BigNumber from 'bignumber.js';
import { fetchPricesAction } from '../actions/prices';
import { Token } from '../entities/token';

/**
 * State containing price infos indexed by token id
 * We only want to track one price per token, not token implem for now
 */
export interface TokenPriceState {
  // todo: should this be a TokenImplem instead?
  // todo: Do we really need a BigNumber
  [tokenId: Token['id']]: BigNumber;
}
const initialState: TokenPriceState = {};

export const tokenPriceSlice = createSlice({
  name: 'token_price',
  initialState: initialState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    // when prices are changed, update prices
    // this could also just be a a super quick drop in replacement
    // if we are OK to not use BigNumber, which I don't think we are
    builder.addCase(fetchPricesAction.fulfilled, (state, action) => {
      for (const tokenId of Object.keys(action.payload)) {
        const tokenPrice = action.payload[tokenId];
        // new price, add it
        if (state.byTokenId[tokenId] === undefined) {
          state.byTokenId[tokenId] = new BigNumber(tokenPrice);

          // price exists, update it if it changed
        } else if (state.byTokenId[tokenId].comparedTo(tokenPrice) === 0) {
          state.byTokenId[tokenId] = new BigNumber(tokenPrice);
        }
      }
    });
  },
});
