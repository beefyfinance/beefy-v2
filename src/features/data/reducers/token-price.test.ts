import { fetchPricesAction, fetchLPPricesAction } from '../actions/prices';
import { BeefyAPITokenPricesResponse } from '../apis/beefy';
import { tokenPriceSlice, initialTokenPriceState } from './token-price';

describe('Token price slice tests', () => {
  it('should do nothing on pending price list', () => {
    const action = { type: fetchPricesAction.pending };
    const state = tokenPriceSlice.reducer(initialTokenPriceState, action);
    expect(state).toEqual(initialTokenPriceState);
  });

  it('should do nothing on rejected price list', () => {
    const action = { type: fetchPricesAction.rejected };
    const state = tokenPriceSlice.reducer(initialTokenPriceState, action);
    expect(state).toEqual(initialTokenPriceState);
  });

  it('should update state on fulfilled vault list', () => {
    const payload: BeefyAPITokenPricesResponse = {
      '1AAVE': 238.16568740268866,
      '1AXS': 79.04891966147113,
      '1BTC': 609115.1079533856,
    };
    const action = { type: fetchPricesAction.fulfilled, payload: payload };
    const state = tokenPriceSlice.reducer(initialTokenPriceState, action);
    expect(state).toMatchSnapshot();
  });

  it('should do nothing on pending lp price list', () => {
    const action = { type: fetchLPPricesAction.pending };
    const state = tokenPriceSlice.reducer(initialTokenPriceState, action);
    expect(state).toEqual(initialTokenPriceState);
  });

  it('should do nothing on rejected lp price list', () => {
    const action = { type: fetchLPPricesAction.rejected };
    const state = tokenPriceSlice.reducer(initialTokenPriceState, action);
    expect(state).toEqual(initialTokenPriceState);
  });

  it('should update state on fulfilled vault list', () => {
    const payload: BeefyAPITokenPricesResponse = {
      '1inch-1inch-bnb': 0,
      '2omb-2omb-ftm': 16.710317081168196,
      '2omb-2share-ftm': 171.84420463276956,
    };
    const action = { type: fetchLPPricesAction.fulfilled, payload: payload };
    const state = tokenPriceSlice.reducer(initialTokenPriceState, action);
    expect(state).toMatchSnapshot();
  });
});
