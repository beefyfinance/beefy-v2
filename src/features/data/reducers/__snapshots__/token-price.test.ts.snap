// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Token price slice tests should update state on fulfilled lp price list 1`] = `
Object {
  "byTokenId": Object {
    "1inch-1inch-bnb": "0",
    "2omb-2omb-ftm": "16.710317081168196",
    "2omb-2share-ftm": "171.84420463276956",
  },
}
`;

exports[`Token price slice tests should update state on fulfilled vault list 1`] = `
Object {
  "byTokenId": Object {
    "1AAVE": "238.16568740268866",
    "1AXS": "79.04891966147113",
    "1BTC": "609115.1079533856",
  },
}
`;
